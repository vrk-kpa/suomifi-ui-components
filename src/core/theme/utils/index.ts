export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
