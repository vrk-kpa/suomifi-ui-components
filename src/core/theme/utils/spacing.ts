import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

export type GlobalMarginProps = { globalMargins: GlobalMargins };

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export type GlobalMargins = {
  all?: MarginProps;
  alert?: MarginProps;
  actionMenu?: MarginProps;
  block?: MarginProps;
  button?: MarginProps;
  breadcrumb?: MarginProps;
  checkbox?: MarginProps;
  checkboxGroup?: MarginProps;
  chip?: MarginProps;
  dateInput?: MarginProps;
  dropdown?: MarginProps;
  expander?: MarginProps;
  expanderGroup?: MarginProps;
  externalLink?: MarginProps;
  fileInput?: MarginProps;
  heading?: MarginProps;
  hintText?: MarginProps;
  inlineAlert?: MarginProps;
  label?: MarginProps;
  languageMenu?: MarginProps;
  link?: MarginProps;
  linkList?: MarginProps;
  loadingSpinner?: MarginProps;
  multiSelect?: MarginProps;
  notification?: MarginProps;
  pagination?: MarginProps;
  paragraph?: MarginProps;
  radioButton?: MarginProps;
  radioButtonGroup?: MarginProps;
  routerLink?: MarginProps;
  searchInput?: MarginProps;
  serviceNavigation?: MarginProps;
  sideNavigation?: MarginProps;
  singleSelect?: MarginProps;
  staticChip?: MarginProps;
  statusText?: MarginProps;
  text?: MarginProps;
  textarea?: MarginProps;
  textInput?: MarginProps;
  timeInput?: MarginProps;
  toast?: MarginProps;
  toggleInput?: MarginProps;
  toggleButton?: MarginProps;
  tooltip?: MarginProps;
  wizardNavigation?: MarginProps;
};

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (
  spacing?: SpacingProps,
  important?: boolean,
): string => {
  if (!spacing) return '';

  const cssStyles = Object.entries(spacing)
    .map(([key, value]) =>
      getCSSSpacing(
        defaultSuomifiTheme,
        key as keyof SpacingProps,
        value,
        important,
      ),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
  important: boolean = false,
) => {
  const amount = spaceVal(theme)(value);
  const importantValue = important ? '!important' : '';
  switch (key) {
    case 'mx':
      return `margin-right: ${amount} ${importantValue}; margin-left: ${amount} ${importantValue};`;
    case 'my':
      return `margin-top: ${amount} ${importantValue}; margin-bottom: ${amount} ${importantValue};`;
    case 'px':
      return `padding-right: ${amount} ${importantValue}; padding-left: ${amount} ${importantValue};`;
    case 'py':
      return `padding-top: ${amount} ${importantValue}; padding-bottom: ${amount} ${importantValue};`;
    default:
      return `${cssSelector[key]}: ${amount} ${importantValue};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
