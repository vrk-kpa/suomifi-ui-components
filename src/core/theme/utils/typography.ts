import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
