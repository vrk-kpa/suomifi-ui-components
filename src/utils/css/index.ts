export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
