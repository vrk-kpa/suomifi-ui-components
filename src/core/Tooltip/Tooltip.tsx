import React, { Component, ReactNode, forwardRef, createRef } from 'react';
import classnames from 'classnames';
import { TooltipContent } from './TooltipContent/TooltipContent';
import { TooltipToggleButton } from './TooltipToggleButton/TooltipToggleButton';
import { forkRefs } from '../../utils/common';
import { MarginProps } from '../theme/utils/spacing';
import { HtmlButtonProps } from '../../reset';

const baseClassName = 'fi-tooltip';

export interface TooltipProps extends MarginProps, HtmlButtonProps {
  /** Tooltip content */
  children: ReactNode;
  /** Toggle button label for screen readers. Should provide context as to which input the Tooltip's information is attached to.
   * Example: If the input is 'First name', `ariaToggleButtonLabelText` would be something like 'First name, show additional information'
   */
  ariaToggleButtonLabelText: string;
  /** Close button label for screen readers. Should provide context as to which input the Tooltip's information is attached to.
   * Example: If the input is 'First name', `ariaCloseButtonLabelText` would be something like 'First name, hide additional information'
   */
  ariaCloseButtonLabelText: string;
  /** Controlled tooltip open state */
  open?: boolean;
  /** CSS class for the Tooltip toggle button */
  toggleButtonClassName?: string;
  /** CSS class for the Tooltip content */
  contentClassName?: string | undefined;
  /**
   * Anchor element for listening to resize events.
   * Used for triggering tooltip content arrow indicator reposition.
   * Should preferably be the first resizable parent container, unless there is none before document body.
   * NOTE: Does not work with inline elements as those do not support resize events.
   */
  anchorElement?: HTMLElement | null;
  /** Callback fired on toggle button click */
  onToggleButtonClick?: (event: React.MouseEvent) => void;
  /** Callback fired on close button click */
  onCloseButtonClick?: (event: React.MouseEvent) => void;
  /** Ref object is forwarded to the toggle button element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLButtonElement>;
}

type TooltipState = {
  open: boolean;
  contentArrowOffsetPx: number;
  anchorElement: HTMLElement | null;
  anchorRefObserver: ResizeObserver | null;
};

class BaseTooltip extends Component<TooltipProps & { className?: string }> {
  state: TooltipState = {
    open: false,
    contentArrowOffsetPx: 0,
    anchorElement: null,
    anchorRefObserver: null,
  };

  private toggleButtonRef: React.RefObject<HTMLButtonElement>;

  private contentRef: React.RefObject<HTMLDivElement>;

  constructor(props: TooltipProps) {
    super(props);
    this.toggleButtonRef = createRef();
    this.contentRef = createRef();
  }

  private calculateContentArrowOffset = () => {
    if (
      !!this.toggleButtonRef &&
      this.toggleButtonRef.current &&
      !!this.contentRef &&
      !!this.contentRef.current
    ) {
      /**
       * Prevent content arrow posisition outside the tooltip content box
       * The arrow is 16 px wide and has a shadow => max pos is 17px less
       */
      const pos =
        this.toggleButtonRef.current.getBoundingClientRect().left -
        this.contentRef.current.getBoundingClientRect().left -
        2;
      const max = this.contentRef.current.getBoundingClientRect().width - 17;
      return Math.round(Math.min(pos, max));
    }
    return 0;
  };

  private setContentArrowOffset = () => {
    this.setState({ contentArrowOffsetPx: this.calculateContentArrowOffset() });
  };

  componentDidMount() {
    window.addEventListener('resize', this.setContentArrowOffset);
    const anchorRefObserver = new ResizeObserver(() => {
      this.setContentArrowOffset();
    });
    this.setState({ anchorRefObserver });
  }

  componentDidUpdate(prevProps: TooltipProps) {
    if (this.props.anchorElement !== this.state.anchorElement) {
      if (!!this.state.anchorRefObserver) {
        if (!!this.state.anchorElement) {
          this.state.anchorRefObserver.unobserve(this.state.anchorElement);
        }
        if (!!this.props.anchorElement) {
          this.state.anchorRefObserver.observe(this.props.anchorElement);
        }
      }
      this.setState({ anchorElement: this.props.anchorElement });
      this.setContentArrowOffset();
    } else if (prevProps.open !== this.props.open && this.props.open) {
      this.setContentArrowOffset();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.setContentArrowOffset);
    if (!!this.state.anchorRefObserver) {
      this.state.anchorRefObserver.disconnect();
    }
  }

  private handleToggleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    if (!!this.props.onToggleButtonClick) {
      this.props.onToggleButtonClick(event);
    }
    if (!('open' in this.props)) {
      this.setState((prevState: TooltipState) => ({ open: !prevState.open }));
    }
  };

  private handleCloseClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    if (!!this.props.onCloseButtonClick) {
      this.props.onCloseButtonClick(event);
    }
    if (!('open' in this.props)) {
      this.setState({ open: false });
    }
    if (!!this.toggleButtonRef.current) {
      this.toggleButtonRef.current.focus();
    }
  };

  render() {
    const {
      open: propsOpen,
      children,
      ariaToggleButtonLabelText,
      ariaCloseButtonLabelText,
      toggleButtonClassName,
      contentClassName,
      forwardedRef,
      anchorElement,
      className,
      onToggleButtonClick,
      onCloseButtonClick,
      ...passProps
    } = this.props;
    const open = 'open' in this.props ? propsOpen : this.state.open;
    // Remove the possibility to have undefined forwardedRef as a parameter for forkRefs
    const definedRef = forwardedRef || null;

    return (
      <>
        <TooltipToggleButton
          className={classnames(
            className,
            baseClassName,
            toggleButtonClassName,
          )}
          ref={forkRefs(this.toggleButtonRef, definedRef)}
          aria-label={ariaToggleButtonLabelText}
          aria-expanded={open}
          onClick={this.handleToggleClick}
          {...passProps}
        />
        {!!open && (
          <TooltipContent
            arrowOffsetPx={this.state.contentArrowOffsetPx}
            ref={this.contentRef}
            className={contentClassName}
            onCloseButtonClick={this.handleCloseClick}
            ariaCloseButtonLabelText={ariaCloseButtonLabelText}
          >
            {children}
          </TooltipContent>
        )}
      </>
    );
  }
}

const Tooltip = forwardRef(
  (props: TooltipProps, ref: React.RefObject<HTMLButtonElement>) => {
    const { ...passProps } = props;
    return <BaseTooltip forwardedRef={ref} {...passProps} />;
  },
);

Tooltip.displayName = 'Tooltip';
export { Tooltip };
