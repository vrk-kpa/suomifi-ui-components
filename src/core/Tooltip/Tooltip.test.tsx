import React from 'react';
import { render, screen } from '@testing-library/react';
import { Tooltip } from './Tooltip';
import { axeTest } from '../../utils/test';

describe('props', () => {
  /**
 * Note for the following tests:
      For some reason, two identical buttons are returned in the start when there should be only one toggle button.
      Therefore need to use getAllByRole to get the toggle button.
      `const toggleButton = screen.getAllByRole('button')[0];`
 */
  describe('children', () => {
    it('should have the given text', () => {
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
        >
          Children of the component
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      toggleButton.click();
      expect(screen.getByText('Children of the component')).toBeInTheDocument();
    });
  });

  describe('ariaToggleButtonLabelText', () => {
    it('should have the given value', () => {
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      expect(toggleButton).toHaveAttribute('aria-label', 'Toggle tooltip');
    });
  });

  describe('ariaCloseButtonLabelText', () => {
    it('should have the given value', () => {
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      toggleButton.click();
      const closeButton = screen.getAllByRole('button')[1];
      expect(closeButton).toHaveAttribute('aria-label', 'Close tooltip');
    });
  });

  describe('toggleButtonClassName', () => {
    test('toggleButton should have the given className', () => {
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
          toggleButtonClassName="custom-class"
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      expect(toggleButton).toHaveClass('custom-class');
    });
  });

  describe('contentClassName', () => {
    test('content should have the given className', () => {
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
          contentClassName="custom-class"
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      toggleButton.click();
      const contentDiv = screen.getAllByRole('button')[1].parentElement;
      expect(contentDiv).toHaveClass('custom-class');
    });
  });

  describe('onToggleButtonClick', () => {
    it('is clicked; given method called', () => {
      const mockClickHandler = jest.fn();
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
          onToggleButtonClick={mockClickHandler}
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      toggleButton.click();
      expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });
  });

  describe('onCloseButtonClick', () => {
    it('is clicked; given method called', () => {
      const mockClickHandler = jest.fn();
      render(
        <Tooltip
          ariaCloseButtonLabelText="Close tooltip"
          ariaToggleButtonLabelText="Toggle tooltip"
          onCloseButtonClick={mockClickHandler}
        >
          Test Tooltip
        </Tooltip>,
      );
      const toggleButton = screen.getAllByRole('button')[0];
      toggleButton.click();
      const closeButton = screen.getAllByRole('button')[1];
      closeButton.click();
      expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });
  });
});

describe('accessibility', () => {
  const TestTooltip = (
    <Tooltip
      ariaCloseButtonLabelText="Close tooltip"
      ariaToggleButtonLabelText="Toggle tooltip"
    >
      Test Tooltip
    </Tooltip>
  );
  test('should not have basic accessibility issues', axeTest(TestTooltip));
});

describe('Basic tooltip', () => {
  it('should match snapshot', () => {
    const BasicTooltip = (
      <Tooltip
        ariaCloseButtonLabelText="Close tooltip"
        ariaToggleButtonLabelText="Toggle tooltip"
      >
        Test Tooltip
      </Tooltip>
    );
    const { container } = render(BasicTooltip);
    const toggleButton = screen.getAllByRole('button')[0];
    toggleButton.click();
    expect(container).toMatchSnapshot();
  });
});

describe('margin', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(
      <Tooltip
        ariaToggleButtonLabelText=""
        ariaCloseButtonLabelText=""
        margin="xs"
      >
        Test
      </Tooltip>,
    );
    expect(container).toMatchSnapshot();
  });
});
