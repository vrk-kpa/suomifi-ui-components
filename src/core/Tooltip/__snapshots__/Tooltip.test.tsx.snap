// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tooltip should match snapshot 1`] = `
.c5 {
  vertical-align: baseline;
}

.c5.fi-icon {
  display: inline-block;
}

.c5 .fi-icon-base-fill {
  fill: currentColor;
}

.c5 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c5.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c5.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c2 {
  vertical-align: baseline;
}

.c2.fi-icon {
  display: inline-block;
}

.c2 .fi-icon-base-fill {
  fill: currentColor;
}

.c2 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c2.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c2.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c0::-webkit-inner-spin-button {
  height: auto;
}

.c0::-webkit-outer-spin-button {
  height: auto;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c4 {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c4.fi-tooltip_content {
  margin: 10px 0;
  position: relative;
  border: 1px solid hsl(201,7%,46%);
  border-radius: 2px;
  background-color: hsl(212,63%,98%);
  padding: 20px 43px 20px 20px;
  box-shadow: 0 2px 3px 0 hsla(0,0%,13%,0.2);
}

.c4.fi-tooltip_content:before,
.c4.fi-tooltip_content:after {
  content: '';
  position: absolute;
  height: 0;
  width: 0;
  border: solid transparent;
  pointer-events: none;
}

.c4.fi-tooltip_content:before {
  border-bottom-color: hsl(0,0%,13%);
  border-width: 9px;
  left: 0px;
  margin-right: -9px;
  bottom: 100%;
}

.c4.fi-tooltip_content:after {
  border-bottom-color: hsl(212,63%,98%);
  border-width: 8px;
  margin-right: -9px;
  bottom: 100%;
  left: calc(0px + 1px);
}

.c4.fi-tooltip_content.fi-tooltip_content--arrow-hidden:before,
.c4.fi-tooltip_content.fi-tooltip_content--arrow-hidden:after {
  display: none;
}

.c4.fi-tooltip_content .fi-tooltip_close-button {
  position: absolute;
  top: 0px;
  right: 0px;
  height: 40px;
  width: 40px;
  padding: 12px;
  border-radius: 2px;
}

.c4.fi-tooltip_content .fi-tooltip_close-button .fi-tooltip_close-button_icon {
  width: 16px;
  height: 16px;
}

.c4.fi-tooltip_content .fi-tooltip_close-button:active {
  background: linear-gradient(-180deg,hsl(202,7%,80%) 0%,hsla(0,0%,100%,0) 100%);
}

.c4.fi-tooltip_content .fi-tooltip_close-button:focus-visible {
  outline: 3px solid transparent;
}

.c4.fi-tooltip_content .fi-tooltip_close-button:focus-visible:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c4.fi-tooltip_content .fi-tooltip_close-button:hover {
  outline: 1px solid hsl(0,0%,13%);
}

.c4.fi-tooltip_content .fi-heading {
  margin-bottom: 10px;
  margin-top: 0;
}

.c4.fi-tooltip_content .fi-text {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 {
  display: inline;
  height: 16px;
  width: 16px;
  margin-left: 5px;
  vertical-align: middle;
}

.c1:focus-visible {
  outline: 3px solid transparent;
  position: relative;
}

.c1:focus-visible:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1 .fi-tooltip_toggle-button_icon {
  height: 100%;
  width: 100%;
  color: hsl(212,63%,45%);
}

<div>
  <button
    aria-expanded="true"
    aria-label="Toggle tooltip"
    class="c0 c1 fi-tooltip fi-tooltip_toggle-button"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="fi-icon c2 fi-tooltip_toggle-button_icon"
      focusable="false"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class="fi-icon-base-fill"
        d="M11.996 0c3.207 0 6.22 1.248 8.487 3.514A11.917 11.917 0 0 1 24 12a11.92 11.92 0 0 1-3.517 8.486A11.927 11.927 0 0 1 11.996 24c-3.207 0-6.221-1.249-8.487-3.514-4.679-4.68-4.679-12.292 0-16.972A11.923 11.923 0 0 1 11.996 0Zm.007 10H10a1 1 0 0 0 0 2h1v5h-1a1 1 0 0 0 0 2h4.003a1 1 0 0 0 0-2h-1v-6a1 1 0 0 0-1-1Zm-.501-5a1.5 1.5 0 1 0-.002 3h.002l.145-.006A1.5 1.5 0 0 0 11.502 5Z"
        fill="#222"
        fill-rule="evenodd"
      />
    </svg>
  </button>
  <div
    class="c3 fi-tooltip_content c4 fi-tooltip_content--arrow-hidden"
  >
    Test Tooltip
    <button
      aria-label="Close tooltip"
      class="c0 fi-tooltip_close-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="fi-icon c5 fi-tooltip_close-button_icon"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="fi-icon-base-fill"
          d="M4.75 3.3 12 10.55l7.25-7.25a1.026 1.026 0 0 1 1.45 1.45L13.45 12l7.25 7.25a1.026 1.026 0 0 1-1.45 1.45L12 13.45 4.75 20.7a1.026 1.026 0 0 1-1.45-1.45L10.55 12 3.3 4.75A1.026 1.026 0 0 1 4.75 3.3Z"
          fill="#222"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`margin should have margin style from margin prop 1`] = `
.c2 {
  vertical-align: baseline;
}

.c2.fi-icon {
  display: inline-block;
}

.c2 .fi-icon-base-fill {
  fill: currentColor;
}

.c2 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c2.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c2.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c0::-webkit-inner-spin-button {
  height: auto;
}

.c0::-webkit-outer-spin-button {
  height: auto;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  display: inline;
  height: 16px;
  width: 16px;
  margin-left: 5px;
  vertical-align: middle;
}

.c1:focus-visible {
  outline: 3px solid transparent;
  position: relative;
}

.c1:focus-visible:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1 .fi-tooltip_toggle-button_icon {
  height: 100%;
  width: 100%;
  color: hsl(212,63%,45%);
}

<div>
  <button
    aria-expanded="false"
    aria-label=""
    class="c0 c1 fi-tooltip fi-tooltip_toggle-button"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="fi-icon c2 fi-tooltip_toggle-button_icon"
      focusable="false"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class="fi-icon-base-fill"
        d="M11.996 0c3.207 0 6.22 1.248 8.487 3.514A11.917 11.917 0 0 1 24 12a11.92 11.92 0 0 1-3.517 8.486A11.927 11.927 0 0 1 11.996 24c-3.207 0-6.221-1.249-8.487-3.514-4.679-4.68-4.679-12.292 0-16.972A11.923 11.923 0 0 1 11.996 0Zm.007 10H10a1 1 0 0 0 0 2h1v5h-1a1 1 0 0 0 0 2h4.003a1 1 0 0 0 0-2h-1v-6a1 1 0 0 0-1-1Zm-.501-5a1.5 1.5 0 1 0-.002 3h.002l.145-.006A1.5 1.5 0 0 0 11.502 5Z"
        fill="#222"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;
